saga/saga.go

```diff
@@ -1,146 +1,197 @@
- * has been avoided and in Rollback Recovery Mode. Should not be used in ForwardRecovery Mode
- * returns an error if it fails
- *
- * StartCompTask is idempotent with respect to sagaId & taskId.  If
- * the data passed changes the last written StartCompTask message will win
- *
- * Returns the resulting SagaState or an error if it fails
- */
-func (s Saga) StartCompensatingTask(state *SagaState, taskId string, data []byte) (*SagaState, error) {
-	return s.logMessage(state, MakeStartCompTaskMessage(state.sagaId, taskId, data))
+//
+// Log a Start Compensating Task Message to the log. Should only be logged after a Saga
+// has been avoided and in Rollback Recovery Mode. Should not be used in ForwardRecovery Mode
+// returns an error if it fails
+//
+// StartCompTask is idempotent with respect to sagaId & taskId.  If
+// the data passed changes the last written StartCompTask message will win
+//
+// Returns an error if it fails
+//
+func (s *Saga) StartCompensatingTask(taskId string, data []byte) error {
+	return s.updateSagaState(MakeStartCompTaskMessage(s.id, taskId, data))
 }
 
-/*
- * Log an End Compensating Task Message to the log when a Compensating Task
- * has been successfully completed. Returns an error if it fails.
- *
- * EndCompTask is idempotent with respect to sagaId & taskId.  If
- * the data passed changes the last written EndCompTask message will win
- *
- * Returns the resulting SagaState or an error if it fails
- */
-func (s Saga) EndCompensatingTask(state *SagaState, taskId string, results []byte) (*SagaState, error) {
-	return s.logMessage(state, MakeEndCompTaskMessage(state.sagaId, taskId, results))
+//
+// Log an End Compensating Task Message to the log when a Compensating Task
+// has been successfully completed. Returns an error if it fails.
+//
+// EndCompTask is idempotent with respect to sagaId & taskId.  If
+// the data passed changes the last written EndCompTask message will win
+//
+// Returns an error if it fails
+//
+func (s *Saga) EndCompensatingTask(taskId string, results []byte) error {
+	return s.updateSagaState(MakeEndCompTaskMessage(s.id, taskId, results))
 }
 
-/*
- * Should be called at Saga Creation time.
- * Returns a Slice of In Progress SagaIds
- */
-func (s Saga) Startup() ([]string, error) {
+// updateSagaStateLoop that is executed inside of a single go routine.  There
+// is one per saga currently executing.  This ensures all updates are applied
+// in order to a saga.  Also controls access to the SagaState so its is
+// updated in a thread safe manner
+func (s *Saga) updateSagaStateLoop() {
+	for update := range s.updateCh {
+		s.mutex.Lock()
+		currState := s.state
+		newState, err := logMessage(currState, update.msg, s.log)
+
+		if err == nil {
+			s.state = newState
+
+			if update.msg.msgType == EndSaga {
+				close(s.updateCh)
+			}
+
+			update.resultCh <- nil
+
+		} else {
+			update.resultCh <- err
+		}
+		s.mutex.Unlock()
```

Comment:
Oh, duh. Yeah.

I'd probably refactor to make each iteration just call a function, that then is able to defer the unlock. This will save us a deadlock if, e.g., logMessage panics.

But it's also more code and functions, so I understand if you'd prefer not to.